eng/scripts/live-test-resource-cleanup.ps1

```diff
@@ -3,93 +3,373 @@
+      return $ResourceGroup.Tags[$tagKey]
+    }
+  }
 
-$hasDeleteAfter = $allGroups.Where({ $_.Tags.Keys -contains "DeleteAfter" })
-Write-Host "Count $($hasDeleteAfter.Count)"
-$toDelete = $hasDeleteAfter.Where({ $deleteDate = ($_.Tags.DeleteAfter -as [DateTime]); (!$deleteDate -or $now -gt $deleteDate) })
-Write-Host "Groups to delete: $($toDelete.Count)"
+  return $null
+}
 
-# Get purgeable resources already in a deleted state.
-$purgeableResources = @(Get-PurgeableResources)
+function HasValidOwnerTag([object]$ResourceGroup) {
+  $ownerTag = GetTag $ResourceGroup "Owners"
+  if (!$ownerTag) {
+    return $false
+  }
+  $owners = $ownerTag -split "[;, ]"
+  $hasValidOwner = $false
+  $invalidOwners = @()
+  foreach ($owner in $owners) {
+    if (IsValidAlias -Alias $owner) {
+      $hasValidOwner = $true
+    } else {
+      $invalidOwners += $owner
+    }
+  }
+  if ($invalidOwners) {
+    Write-Warning " Resource group '$($ResourceGroup.ResourceGroupName)' has invalid owner tags: $($invalidOwners -join ',')"
+  }
+  if ($hasValidOwner) {
+    Write-Host " Skipping tagged resource group '$($ResourceGroup.ResourceGroupName)' with owners '$owners'"
+    return $true
+  }
+  return $false
+}
 
-foreach ($rg in $toDelete)
-{
-  if ($Force -or $PSCmdlet.ShouldProcess("$($rg.ResourceGroupName) (UTC: $($rg.Tags.DeleteAfter))", "Delete Group")) {
-    # Add purgeable resources that will be deleted with the resource group to the collection.
-    $purgeableResourcesFromRG = Get-PurgeableGroupResources $rg.ResourceGroupName
+function HasValidAliasInName([object]$ResourceGroup) {
+    # check compliance (formatting first, then validate alias) and skip if compliant
+    if ($ResourceGroup.ResourceGroupName `
+      -match '^(rg-)?(?<alias>(t-|a-|v-)?[a-z,A-Z]+)([-_]{1}.*)?$' `
+      -and (IsValidAlias -Alias $matches['alias']))
+    {
+      Write-Host " Skipping resource group '$($ResourceGroup.ResourceGroupName)' starting with valid alias '$($matches['alias'])'"
+      return $true
+    }
+    return $false
+}
 
-    if ($purgeableResourcesFromRG) {
-      $purgeableResources += $purgeableResourcesFromRG
-      Write-Verbose "Found $($purgeableResourcesFromRG.Count) potentially purgeable resources in resource group $($rg.ResourceGroupName)"
+function GetDeleteAfterTag([object]$ResourceGroup) {
+  return GetTag $ResourceGroup "DeleteAfter"
+}
+
+function HasExpiredDeleteAfterTag([string]$DeleteAfter) {
+  if ($DeleteAfter) {
+    $deleteDate = $deleteAfter -as [DateTime]
+    return $deleteDate -and [datetime]::UtcNow -gt $deleteDate
+  }
+  return $false
+}
+
+function HasException([object]$ResourceGroup) {
+    if ($Exceptions.Count) {
+      if ($Exceptions.Contains($ResourceGroup.ResourceGroupName)) {
+        Write-Host " Skipping allowed resource group '$($ResourceGroup.ResourceGroupName)' because it is in the allow list '$AllowListPath'"
+        return $true
+      }
+      return $false
+    }
+    if ($ShouldCheckAllowList) {
```

Comment:
Who go through this extra level of checking? You should just initialize the exception set and if there is nothing in there then the contains check will return very quickly.